plugins/redis/plugin.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2018-present the CoreDHCP Authors. All rights reserved
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+package redisplugin
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/coredhcp/coredhcp/handler"
+	"github.com/coredhcp/coredhcp/logger"
+	"github.com/coredhcp/coredhcp/plugins"
+	"github.com/gomodule/redigo/redis"
+	"github.com/insomniacslk/dhcp/dhcpv4"
+	"github.com/insomniacslk/dhcp/dhcpv6"
+)
+
+// various global variables
+var log = logger.GetLogger("plugins/redis")
+var pool *redis.Pool
+var leaseTime time.Duration
+
+func init() {
+	plugins.RegisterPlugin("redis", setupRedis6, setupRedis4)
+}
+
+// Handler6 handles DHCPv6 packets for the redis plugin
+func Handler6(req, resp dhcpv6.DHCPv6) (dhcpv6.DHCPv6, bool) {
+	// TODO add IPv6 support
+	return nil, true
+}
+
+// Handler4 handles DHCPv4 packets for the redis plugin
+func Handler4(req, resp *dhcpv4.DHCPv4) (*dhcpv4.DHCPv4, bool) {
+	// Get redis connection from pool
+	conn := pool.Get()
+
+	defer conn.Close()
+
+	// Get all options for a MAC
+	options, err := redis.StringMap(conn.Do("HGETALL", "mac:"+req.ClientHWAddr.String()))
+	if err == redis.ErrNil || options["ipv4"] == "" {
+		log.Printf("MAC address %s is not allowed", req.ClientHWAddr.String())
+		return nil, true
+	} else if err != nil {
+		log.Printf("Redis error: %s", err)
+		return nil, true
+	}
+
+	// Loop through options returned and assign as needed
+	for option, value := range options {
+		switch option {
+		case "ipv4":
+			ipaddr, ipnet, err := net.ParseCIDR(value)
+			if err != nil {
+				log.Printf("malformed IP %s error: %s", value, err)
+				return nil, true
+			}
+			resp.YourIPAddr = ipaddr
+			resp.Options.Update(dhcpv4.OptSubnetMask(ipnet.Mask))
+			log.Printf("found IP address %s for MAC %s", value, req.ClientHWAddr.String())
+
+		case "router":
+			router := net.ParseIP(value)
+			if router.To4() == nil {
+				log.Printf("Invalid router option: %s for MAC %s", value, req.ClientHWAddr)
+			}
+			resp.Options.Update(dhcpv4.OptRouter(router))
+
+		case "dns":
+			var dnsServers4 []net.IP
+			servers := strings.Split(value, ",")
+			for _, server := range servers {
+				DNSServer := net.ParseIP(server)
+				if DNSServer.To4() == nil {
+					log.Printf("Invalid dns server: %s for MAC %s", server, req.ClientHWAddr)
+					return nil, true
+				}
+				dnsServers4 = append(dnsServers4, DNSServer)
+			}
+			if req.IsOptionRequested(dhcpv4.OptionDomainNameServer) {
+				resp.Options.Update(dhcpv4.OptDNS(dnsServers4...))
+			}
+
+		default:
+			log.Printf("found unhandled option %s for MAC %s", option, req.ClientHWAddr.String())
+		}
+	}
+
+	// Set lease time
+	resp.Options.Update(dhcpv4.OptIPAddressLeaseTime(leaseTime))
+
+	// If request has Relay Agent Info copy it to the reply
```

Comment:
This is not specific to the redis plugin, I think it should be moved to the main handler:
https://github.com/coredhcp/coredhcp/blob/master/coredhcp.go#L169

Reply:
I was actually thinking that I should add support for a redis field for lease time and if it exists then update it.   Perhaps there could be a default in the main handler and then the redis plugin over-ride it.  What do you think about that?

Reply:
I was referring to the Relay Agent Info not the lease time. I wouldn't place the lease time in the main handler. Maybe we can have a lease time plugin and the redis plugin over-ride it.

Reply:
ahhh...sorry...yes the relay agent info is definitely not specific and to be honest I'm not 100% sure it has to be copied according to the RFCs but from wireshark captures the ISC servers copy it.  For the lease time I think the current way of the plugin having a lease time option set in the config file works fine and I could add it to allow redis to override in that plugin.  Do you think I should add support for more options in the redis plugin or just add them when/if anyone else needed them?

Reply:
Do you want me to make an attempt at moving the Relay Agent Info stuff into the main handler in this pull request?

Reply:
I think it should go either into the dhcp library instead (github.com/insomniacslk/dhcp), to be included in `dhcpv4.NewReplyFromRequest`; Or in a separate plugin.

The RFC says "DHCP servers claiming to support the Relay Agent Information option SHALL echo the entire contents of the Relay Agent Information option in all replies", but we don't have to support it, so it's fine to just drop it from this

Reply:
It seems like adding another plugin would not be the best solution if it should be copied if it exists...that could potentially lead to unexpected behavior for a new user switch from another server and missing that they need to enable a plugin.  The dhcp library NewReplyFromRequest seems to be the most logical place for it.  What do you think @insomniacslk ?

Reply:
Yes, maybe the library is a better place for it. I have seen some relays dropping the packet if the option in not there.

Reply:
It definitely should not be added here. We accept a library PR :) Otherwise, I can add it.

Reply:
I can take a stab at the library PR later today.